### PR TITLE
Lifetime cleanups for `basic_string`

### PR DESCRIPTION
**Pull request by achabense**
Resolves the comments [(1)](https://github.com/microsoft/STL/pull/4031#discussion_r1331063123) [(2)](https://github.com/microsoft/STL/pull/4031#issuecomment-1727571761) in pr 4031. Hope that is the last lifetime bug of `basic_string` using fancy pointers (`3.1.`).
<hr>

Sometimes the pointer's destruction and buffer's activation are dangerously separated into different functions, in an obscure way (mainly relevant to the usage of `_Tidy_init`).
This pr introduces a helper function `_String_val::_Bxty::_Switch_to_buf` to make the transition state (where both `_Ptr` and `_Buf` are inactive) invisible to `basic_string`'s ctors and methods.
<hr>

Remaining issue after this pr:
https://github.com/microsoft/STL/blob/48eedd369d6c88130fff4d15afb88f016fa6e648/stl/inc/xstring#L2282-L2284
As `_String_val` already initializes the buffer on construction, and every switch back to small mode is now protected by the helper function, in `basic_string`'s ctor and methods, when we can decide a string is in small mode, we can also assume that elements in `_Buf` are activated. *Therefore, every remaining usage of `_Activate_SSO_buffer` (including the one in newly added `_Construct_empty`) becomes redundant*...
Or, `Don't rely on the new behavior`? Does this mean we should not assume that `_Buf` is initialized in ctor, even if it does so now? Can we remove the remaining `_Activate_SSO_buffer` then? I'm highly horrified by this; especially, what's its relation with the lifetime issue of fancy pointers?

Related: #1735 #3235 and #3273.

### CODE REVIEW

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:4843]: Slightly irrelevant; it also seems that the block in this function is actually useless (`_Tidy_deallocate` seems always followed by methods that just rewrite these values).
```
        _My_data._Mysize = 0;
        _My_data._Myres  = _Small_string_capacity;
        // the _Traits::assign is last so the codegen doesn't think the char write can alias this
        _Traits::assign(_My_data._Bx._Buf[0], _Elem());
```

File: tests/std/tests/VSO_0000000_allocator_propagation/test.cpp
Comment (achabense) [inline tests/std/tests/VSO_0000000_allocator_propagation/test.cpp:949]: This don't have to be guarded by `_HAS_CXX20`, as `strbuf(move(str))` will still call copy constructor in pre-cpp20 mode. However, I think adding `_HAS_CXX20` will make it more obvious that the leak is relevant to `basic_stringbuf(basic_string&&,...)` introduced in C++20.

**Comment by achabense** [comment]
https://github.com/microsoft/STL/blob/48eedd369d6c88130fff4d15afb88f016fa6e648/stl/inc/xstring#L2282-L2284
`Don't rely on the new behavior` Does this mean we should not assume that `_Buf` is initialized on ctor, even if it does so?

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring]: I'm not too sure whether it's good to add `noexcept` for this function.

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:2644]: I'm not too sure whether it's good to add `noexcept` for this function.

**Comment by StephanTLavavej** [comment]
Regarding the "Don't rely on the new behavior" comment - if you look at `git blame`, this was introduced by #3235 (which you noticed was related), because we found an ODR mix-and-match problem between 17.3 and 17.4. The problem was a case of "don't make coordinated behavioral changes between two functions" because the linker might mix old `a()` with new `b()` or new `a()` with old `b()`. This is pretty rare for us, so we aren't good at noticing when it's about to happen (unlike representation changes).

I'm pretty nervous about changes in this area as a result. The idea behind `_String_val::_Bxty::_Switch_to_buf` appears highly desirable, but we need to be careful about assuming that another member function (or a constructor) has left the object in a known state, if that could have been different in 17.3 and earlier.

It might be time to rename `_String_val` now, even though that will make this PR bigger, to reduce the risk of introducing a damaged state. We'll still need to maintain layout compatibility (and keep the visualizers working), but we won't need to worry about mix-and-match behavior. (The linker could still pick old and new `basic_string` member functions, but those all leave the object in a guaranteed state.)

File: stl/inc/xstring
Comment (StephanTLavavej) [reply stl/inc/xstring:2644]: It doesn't *need* to be noexcept - any (non-conformant) exceptions thrown by the proxy allocation will slam into the `basic_string` constructor's annotation. Arguably, this shouldn't actually be marked - note that `basic_string() noexcept(is_nothrow_default_constructible_v<_Alty>)` permits exceptions to be emitted (although this is from a surprising source). I think we should remove this marking and the comment.

Comment (StephanTLavavej) [reply stl/inc/xstring:2644]: It's also a behavioral change compared to the previous direct allocation of a proxy.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for these fine-grained commits, they made reviewing very straightforward! After thinking about each one, I believe that this PR has no ABI impact, and we should proceed as-is since it fixes the lifetime bug and generally makes it harder to introduce such mistakes in the future. Each commit is a transformation that doesn't introduce new inter-function assumptions (as far as I can tell).

This PR is definitely complicated enough and changing a highly-used, risky area of the codebase that we'll need a second maintainer signoff.

It definitely seems like any followup PR to remove `_Activate_SSO_buffer` needs to also rename `_String_val`, otherwise it'll be making the same ABI break as we suffered previously.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for making `basic_string` more maintainable! :tada: :heart_eyes_cat: :yarn:

**Pull request merged**
